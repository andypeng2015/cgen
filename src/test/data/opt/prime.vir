module foo

export function w $foo(w %n) {
@start:
  %c = le.w %n, 1_w
  br %c, @no, @twothree
@no:
  ret 0_w
@twothree:
  %c = le.w %n, 3_w
  br %c, @yes, @divtwo
@yes:
  ret 1_w
@divtwo:
  %r = urem.w %n, 2_w
  %c = eq.w %r, 0_w
  br %c, @yes, @divthree
@divthree:
  %r = urem.w %n, 3_w
  %c = eq.w %r, 0_w
  br %c, @yes, @main
@main:
  %i = copy.w 5_w
  jmp @loop
@loop:
  %s = mul.w %i, %i
  %c = le.w %s, %n
  br %c, @divi, @yes
@divi:
  %r = urem.w %n, %i
  %c = eq.w %r, 0_w
  br %c, @no, @divi2
@divi2:
  %i2 = add.w %i, 2_w
  %r = urem.w %n, %i2
  %c = eq.w %r, 0_w
  br %c, @no, @inc
@inc:
  %i = add.w %i, 6_w
  jmp @loop
}
