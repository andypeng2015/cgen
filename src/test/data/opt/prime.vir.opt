module foo

export function w $foo(w %n) {
@2:
  %16 = le.w %n, 0x1_w ; @44
  br %16, @28(0x0_w), @4
@4:
  %17 = le.w %n, 0x3_w ; @45
  br %17, @28(0x1_w), @7
@7:
  %18 = and.w %n, 0x1_w ; @46
  %19 = eq.w %18, 0x0_w ; @47
  br %19, @28(0x0_w), @9
@9:
  %20 = umulh.w %n, 0x55555556_w ; @48
  %21 = lsl.w %20, 0x1_w ; @49
  %22 = add.w %21, %20 ; @50
  %23 = sub.w %n, %22 ; @51
  %24 = eq.w %23, 0x0_w ; @52
  br %24, @28(0x0_w), @15(0x5_w)
@15(%i.2):
  %25 = mul.w %i.2, %i.2 ; @53
  %26 = le.w %25, %n ; @54
  br %26, @17, @28(0x1_w)
@17:
  %r.3 = urem.w %n, %i.2 ; @21
  %27 = eq.w %r.3, 0x0_w ; @55
  br %27, @28(0x0_w), @20
@20:
  %28 = add.w %i.2, 0x2_w ; @56
  %r.4 = urem.w %n, %28 ; @25
  %29 = eq.w %r.4, 0x0_w ; @57
  br %29, @28(0x0_w), @23
@23:
  %30 = add.w %i.2, 0x6_w ; @58
  jmp @15(%30)
@28(%0):
  ret %0
}
