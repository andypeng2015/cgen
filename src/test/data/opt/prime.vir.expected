module foo

export function w $foo(w %n) {
@24:
  %0 = le.w %n, 0x1_w ; @26
  br %0, @4, @22
@4:
  ret 0x0_w
@22:
  %1 = le.w %n, 0x3_w ; @27
  br %1, @11, @19
@11:
  ret 0x1_w
@19:
  %3 = and.w %n, 0x1_w ; @29
  %2 = eq.w %3, 0x0_w ; @28
  br %2, @11, @16
@16:
  %8 = umulh.w %n, 0x55555556_w ; @34
  %7 = lsl.w %8, 0x1_w ; @33
  %6 = add.w %7, %8 ; @32
  %5 = sub.w %n, %6 ; @31
  %4 = eq.w %5, 0x0_w ; @30
  br %4, @11, @1(0x5_w)
@1(%i.2):
  %10 = mul.w %i.2, %i.2 ; @36
  %9 = le.w %10, %n ; @35
  br %9, @8, @11
@8:
  %r.3 = urem.w %n, %i.2 ; @10
  %11 = eq.w %r.3, 0x0_w ; @37
  br %11, @4, @3
@3:
  %12 = add.w %i.2, 0x2_w ; @38
  %r.4 = urem.w %n, %12 ; @6
  %13 = eq.w %r.4, 0x0_w ; @39
  br %13, @4, @0
@0:
  %14 = add.w %i.2, 0x6_w ; @40
  jmp @1(%14)
}
